File: src/main/java/br/com/model/entities/classes/Fornecedor.java
package br.com.model.entities.classes;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;

import com.fasterxml.jackson.annotation.JsonIgnore;

import br.com.model.entities.classes.Endereco.EnderecoFornecedor;
import br.com.model.entities.classes.Telefone.TelefoneFornecedor;
import br.com.model.entities.interfaces.FornecedorInterface;

@Entity
public class Fornecedor implements FornecedorInterface {
    @Id
    @GeneratedValue
    private Integer id;
    private String nome;
    private String descricao;

    @OneToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "id_endereco_fornecedor")
    private EnderecoFornecedor endereco;

    private String cnpjCpf;
    private Date data;

    @OneToMany(mappedBy = "fornecedor")
    @JsonIgnore
    private List<ProdutoFornecedor> listaProdutos;

    @OneToMany(mappedBy = "fornecedor", cascade = CascadeType.ALL)
    private List<TelefoneFornecedor> listaTelefones;

    public Fornecedor(String nome, String descricao, String cnpjCpf) {
        setNome(nome);
        setDescricao(descricao);
        setEndereco(endereco);
        setCnpjCpf(cnpjCpf);
        setData(new Date());
        setListaProdutoFornecedor(new ArrayList<ProdutoFornecedor>());
        setListaTelefones(new ArrayList<TelefoneFornecedor>());
    }

    public Fornecedor() {

    }

    @Override
    public String toString() {
        return "\n--Fornecedor--\nNome: " + getNome() + "\nCNPJ/CPF: " + getCnpjCpf() + "\nData: " + getData()
                + "\nDescriçao: " + getDescricao() + getEndereco() + "\nID: " + getId() + getListaTelefones()
                + "\nLista de produtos: " + getListaProdutoFornecedor();
    }

    @Override
    public void adicionarProduto(ProdutoFornecedor produtoFornecedor) {
        produtoFornecedor.setFornecedor(this);
        getListaProdutoFornecedor().add(produtoFornecedor);
    }

    @Override
    public void adicionarTelefone(TelefoneFornecedor telefone) {
        getListaTelefones().add(telefone);
    }

    @Override
    public boolean confirmarPedido(int idPedido) {
        // TODO Auto-generated method stub
        return false;
    }

    @Override
    public void removeProduto(ProdutoFornecedor produto) {
        // TODO Auto-generated method stub

    }

    @Override
    public void confirmarEntrega(int idPedido) {
        // TODO Auto-generated method stub

    }

    @Override
    public void atualizarStatusPedido(int idPedido) {
        // TODO Auto-generated method stub

    }

    // Verificar a possibilidade de uma lista de fornecedores.
    @Override
    public Fornecedor buscarFornecedor(String nome) {
        // TODO Auto-generated method stub
        return null;
    }

    @Override
    public boolean confirmarEstoque(int quantidadeEmEstoque) {
        // TODO Auto-generated method stub
        return false;
    }

    // Getters / Setters

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getDescricao() {
        return descricao;
    }

    public void setDescricao(String descricao) {
        this.descricao = descricao;
    }

    public EnderecoFornecedor getEndereco() {
        return endereco;
    }

    public void setEndereco(EnderecoFornecedor endereco) {
        this.endereco = endereco;
    }

    public String getCnpjCpf() {
        return cnpjCpf;
    }

    public void setCnpjCpf(String cnpjCpf) {
        this.cnpjCpf = cnpjCpf;
    }

    public Date getData() {
        return data;
    }

    public void setData(Date data) {
        this.data = data;
    }

    public List<ProdutoFornecedor> getListaProdutoFornecedor() {
        return listaProdutos;
    }

    public void setListaProdutoFornecedor(List<ProdutoFornecedor> listaProdutos) {
        this.listaProdutos = listaProdutos;
    }

    public List<TelefoneFornecedor> getListaTelefones() {
        return listaTelefones;
    }

    public void setListaTelefones(List<TelefoneFornecedor> listaTelefones) {
        this.listaTelefones = listaTelefones;
    }

}

File: src/main/java/br/com/model/usuario/Fornecedor.kt
package br.com.model.usuario

import br.com.model.*
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.OneToMany
import java.util.*

@Entity
data class Fornecedor(
    override var nome: String,
    override var cnpjCpf: String,
    override var senha: String,
    override var email: String,
    override var data: Date = Date(),
    override var tipo: TipoUsuario = TipoUsuario.FORNECEDOR,
    override var enderecos: MutableList<Endereco> = ArrayList(),
    override var telefones: MutableList<Telefone> = ArrayList(),
    override var pedidos: MutableList<Pedido> = ArrayList(),
    var descricao: String? = "Alterar descrição!",
    @field:OneToMany(
        mappedBy = "fornecedor", cascade = [CascadeType.ALL], fetch = FetchType.EAGER
    ) @field:JsonIgnore var listaProdutoFornecedor: MutableList<ProdutoFornecedor> = ArrayList(),
) : Usuario(nome, cnpjCpf, senha, email, data, tipo, enderecos, telefones, pedidos) {

    override fun toString(): String {
        return "\n--Fornecedor-- Nome: ${super.nome} ID: ${super.id} CNPJ/CPF: ${super.cnpjCpf}\nData: ${
            super.data
        }\nDescriçao: $descricao\nEndereços: ${super.enderecos}\nTelefones: ${super.telefones}\nEmail: ${
            super.email
        }\nSenha: ${super.senha}\nLista de Produtos: $listaProdutoFornecedor\nTipo de Usuário: ${super.tipo}"
    }

    fun adicionarProduto(produtoFornecedor: ProdutoFornecedor) {
        produtoFornecedor.fornecedor = this
        listaProdutoFornecedor.add(produtoFornecedor)
    }

    fun removerProduto(produto: ProdutoFornecedor) {
        listaProdutoFornecedor.remove(produto)
    }

    fun confirmarEntrega(idPedido: Int) {
        pedidos.filter { pedido: Pedido -> pedido.id == idPedido }
            .forEach { pedido: Pedido -> pedido.entregue = true }
    }

    fun atualizarStatusPedido(idPedido: Int) {
        // TODO Auto-generated method stub
    }

    fun confirmarEstoque(quantidadePedida: Int, produto: Produto): Boolean {
        var quantidadeTemporaria = 0
        for (produtoFornecedor in listaProdutoFornecedor) {
            if (produtoFornecedor.quantidadeEmEstoque - quantidadePedida > 0 && produtoFornecedor.produto == produto) {
                quantidadeTemporaria = produtoFornecedor.quantidadeEmEstoque
            }
        }
        return quantidadeTemporaria != 0
    }
}
